### PR TITLE
JDK-8310020: MacroAssembler::call_VM(_leaf) doesn't consistently check for conflict with C calling convention.

### PR DESCRIPTION
**Pull request by dafedafe**
# Issue
The `MacroAssembler::call_VM` and `MacroAssembler::call_VM_leaf` methods don't include asserts to check if the passed arguments collide with the C calling convention registers **consistently**. Some versions omit them completely (e.g. aarch64 `MacroAssembler::call_VM_leaf`), other ones use regular asserts (where the `assert_different_registers` function should be used).

# Solution
We use `assert_different_registers` across all `MacroAssembler::call_VM` and `MacroAssembler::call_VM_leaf` where a  check is needed (except for the arm implementation, which doesn't change as it uses fixed registers).

# Testing
This fix includes changes for x86_32/64 and aarch64, which I could test thoroughly but also for **ppc,** **riscv**, and **s390** for which I would **need some help with testing**.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8310020](https://bugs.openjdk.org/browse/JDK-8310020): MacroAssembler::call_VM(_leaf) doesn't consistently check for conflict with C calling convention. (**Enhancement** - P4)

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [841f0181](https://git.openjdk.org/jdk/pull/14470/files/841f01813d3ab4c7f3f801608579f16505e4f92d)
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**)
 * [Martin Doerr](https://openjdk.org/census#mdoerr) (@TheRealMDoerr - **Reviewer**)
 * [Amit Kumar](https://openjdk.org/census#amitkumar) (@offamitkumar - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14470/head:pull/14470` \
`$ git checkout pull/14470`

Update a local copy of the PR: \
`$ git checkout pull/14470` \
`$ git pull https://git.openjdk.org/jdk.git pull/14470/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14470`

View PR using the GUI difftool: \
`$ git pr show -t 14470`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14470.diff">https://git.openjdk.org/jdk/pull/14470.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14470#issuecomment-1592469204)

### CODE REVIEW

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp [hunk 1]
```diff
@@ -1606,31 +1607,27 @@ void MacroAssembler::super_call_VM_leaf(address entry_point, Register arg_0) {
...
   MacroAssembler::call_VM_leaf_base(entry_point, 3);
 }
 
 void MacroAssembler::super_call_VM_leaf(address entry_point, Register arg_0, Register arg_1, Register arg_2, Register arg_3) {
-  assert(arg_0 != c_rarg3, "smashed arg");
-  assert(arg_1 != c_rarg3, "smashed arg");
-  assert(arg_2 != c_rarg3, "smashed arg");
+  assert_different_registers(arg_0, c_rarg1, c_rarg2, c_rarg3);
+  assert_different_registers(arg_1, c_rarg2, c_rarg3);
+  assert_different_registers(arg_2, c_rarg2);
```
Comment (TobiHartmann) [inline src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:1628]: Shouldn't this be

```suggestion
  assert_different_registers(arg_2, c_rarg3);
```
?

Comment (dafedafe) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp]: Totally! Thanks!

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

**Comment by RealFYang** [comment]
Hello, I can help do some test on linux-riscv64 platform.
I thinks we need following extra changes for riscv. Could you please help add this? Thanks.

```
diff --git a/src/hotspot/cpu/riscv/macroAssembler_riscv.cpp b/src/hotspot/cpu/riscv/macroAssembler_riscv.cpp
index e6d1e66e0c0..32a144b2d42 100644
--- a/src/hotspot/cpu/riscv/macroAssembler_riscv.cpp
+++ b/src/hotspot/cpu/riscv/macroAssembler_riscv.cpp
@@ -666,6 +666,7 @@ void MacroAssembler::call_VM_leaf(address entry_point, Register arg_0) {
 }

 void MacroAssembler::call_VM_leaf(address entry_point, Register arg_0, Register arg_1) {
+  assert_different_registers(arg_1, c_rarg0);
   pass_arg0(this, arg_0);
   pass_arg1(this, arg_1);
   call_VM_leaf_base(entry_point, 2);
@@ -673,6 +674,8 @@ void MacroAssembler::call_VM_leaf(address entry_point, Register arg_0, Register

 void MacroAssembler::call_VM_leaf(address entry_point, Register arg_0,
                                   Register arg_1, Register arg_2) {
+  assert_different_registers(arg_1, c_rarg0);
+  assert_different_registers(arg_2, c_rarg0, c_rarg1);
   pass_arg0(this, arg_0);
   pass_arg1(this, arg_1);
   pass_arg2(this, arg_2);
```

**Comment by dafedafe** [comment]
Thanks @RealFYang, I missed it! It should be fixed now.
Thanks for testing too!

**Comment by dafedafe** [comment]
This PR includes changes for ppc and s390 that I couldn't test either.
 @TheRealMDoerr, @offamitkumar could I ask you if you could please run some tests on this change as well for ppc/s390 (and possibly also have a quick look at it)? Thank you very much!

**Review by TheRealMDoerr — APPROVED** [review]
Thanks for cleaning this up! PPC64 part is good.

**Comment by dafedafe** [comment]
@TobiHartmann @offamitkumar @TheRealMDoerr @RealFYang thank you very much for your reviews!

**Pull request closed without merging**